src/pudl/helpers.py

```diff
@@ -1446,6 +1446,67 @@ def dedupe_on_category(
     return dedup_df.drop_duplicates(subset=base_cols, keep="first")
 
 
+def dedupe_and_drop_nas(
+    dedup_df: pd.DataFrame,
+    primary_key_cols: list[str],
+) -> pd.DataFrame:
+    """Deduplicate a df by comparing primary key columns and dropping null rows.
+
+    When a primary key appears twice in a dataframe, and one record is all null other
+    than the primary keys, drop the null row.
+
+    Args:
+        dedup_df: the dataframe with the records to deduplicate.
+        primary_key_cols: list of columns which must not be duplicated.
+
+    Returns:
+        The deduplicated dataframe.
+    """
+    dupes = dedup_df.loc[dedup_df.duplicated(subset=primary_key_cols, keep=False)]
+    dupe_groups = dupes.groupby(primary_key_cols)
+    if (dupe_groups.nunique() > 1).any().any():  # noqa: PD101
+        raise AssertionError(
+            f"Duplicate records with disagreeing data: {dupes[dupes.set_index(primary_key_cols).index.duplicated(keep=False)]}"
+        )
+    deduped = dupe_groups.first().reset_index()
+    # replace the duplicated rows with the deduped versions
+    return pd.concat(
+        [dedup_df.drop_duplicates(subset=primary_key_cols, keep=False), deduped],
+        ignore_index=True,
+    )
+
+
+def standardize_percentages_ratio(
+    frac_df: pd.DataFrame,
+    mixed_cols: list[str],
+) -> pd.DataFrame:
+    """Standardize mixed 0-1 and 0-100 percentage/ratio reporting in a column.
+
+    When a column uses both 0-1 and 0-100 scales to describe percentages, standardize
+    and convert to 0-1 ratios/fractions.
+
+    Args:
+        frac_df: the dataframe with the columns to standardize.
+        mixed_cols: list of columns which should get standardized to the 0-1 scale.
+
+    Returns:
+        The standardized dataframe.
+    """
+    logger.info(f"Standardizing ratios and percentages for {mixed_cols}")
+    for col in mixed_cols:
+        if not pd.api.types.is_numeric_dtype(frac_df[col]):
+            raise AssertionError(
+                f"{col}: Standardization method requires numeric dtype."
+            )
+        frac_df.loc[(frac_df[col] > 1) & (frac_df[col] <= 100), col] /= 100
+        if frac_df[col].max() > 1:
+            logger.warn(
```

Comment:
If it's a warning we'll probably ignore it down the line.

If there are heuristic checks that would identify when we're incorrectly categorizing something as a percentage vs. a fraction, maybe that's a kind of data validation we should be running on any column that's supposed to end up being a fraction?

Besides the case where there's a lot of values around 1% (which we could maybe notice based on there being a bimodal distribution of values ~100x off from each other?) the case where there are a few values that just naturally happen to be less than 1% but which then get multiplied by 100x and thus wildly & systematically magnified could be pretty weird in a different way, and harder to notice, since the values would just get scattered throughout the rest of the spectrum of values between 0.01-1.0

Reply:
Fixed to raise an error and to add a date range for this function to be applied over.